contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -86,23 +27,60 @@ pub fn route_to_router(
     Ok(Response::new().add_messages(router.route(msgs)))
 }
 
-pub fn route_to_nexus(
+pub fn route_message_with_token_to_nexus(
     storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    info: MessageInfo,
+    msg: router_api::Message,
+) -> Result<Response<nexus::execute::Message>> {
+    let token = info
+        .token()
+        .change_context(Error::InvalidToken)?
+        .ok_or(report!(Error::InvalidToken))?;
+
+    route_message_to_nexus(storage, querier, msg, Some(token))?
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Result::Ok)
+}
+
+pub fn route_messages_to_nexus(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
     msgs: Vec<router_api::Message>,
 ) -> Result<Response<nexus::execute::Message>> {
-    let msgs = msgs
-        .into_iter()
-        .filter_map(|msg| match state::is_message_routed(storage, &msg.cc_id) {
-            Ok(true) => None,
-            Ok(false) => Some(Ok(msg)),
-            Err(err) => Some(Err(err)),
-        })
-        .collect::<Result<Vec<_>>>()?;
+    msgs.into_iter()
+        .map(|msg| route_message_to_nexus(storage, querier, msg, None))
+        .collect::<Result<Vec<_>>>()?
+        .then(|msgs| msgs.concat())
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Result::Ok)
+}
 
-    msgs.iter()
-        .try_for_each(|msg| state::set_message_routed(storage, &msg.cc_id))?;
+fn route_message_to_nexus(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    msg: router_api::Message,
+    token: Option<Coin>,
+) -> Result<Vec<CosmosMsg<nexus::execute::Message>>> {
+    if state::is_message_routed(storage, &msg.cc_id)? {
+        return Ok(vec![]);
+    }
+
+    state::set_message_routed(storage, &msg.cc_id)?;
 
-    let msgs: Vec<nexus::execute::Message> = msgs.into_iter().map(Into::into).collect();
+    let client: nexus::Client = client::CosmosClient::new(querier).into();
+    let config = load_config(storage)?;
 
-    Ok(Response::new().add_messages(msgs))
+    let msg: nexus::execute::Message = (msg, token.clone()).into();
+
+    token
+        .into_iter()
+        .map(|token| BankMsg::Send {
+            to_address: config.nexus.to_string(),
+            amount: vec![token],
+        })
+        .map(Into::into)
+        .chain(iter::once(client.route_message(msg)))
```

Comment:
I'm slightly worried about this split into multiple messages, but it's probably fine?
Is there any chance the tokens can be abused through reentry?